cvprac/cvp_api.py

```diff
@@ -3206,4 +3242,470 @@ def workspace_build_status(self, workspace_id, build_id):
+            Args:
+                cc_id (string): The ID for the new change control.
+                name (string): The name for the new change control.
+                tasks (list): A list of Task IDs as strings
+                    Ex: ['10', '11', '12']
+                series (bool): A flag for running tasks in series or
+                    in parallel. Defaults to True for running in series.
+            Returns:
+                response (dict): A dict that contains...
+                Ex: {'value': {'key': {'id':cc_id,
+                      'time': '...'}
+        '''
+        stages = {'values': {'root': {'name': 'root', 'rows': {'values': []}}}}
+        if series == True:
+            for index, task in enumerate(tasks):
+                stage_id = 'stage%d' % index
+                stages['values']['root']['rows']['values'].append({'values':[stage_id]})
+                stages['values'][stage_id] = {'action': {'args': {'values': {'TaskID': task}},'name':'task','timeout':3000},'name':stage_id}
+        else:
+            stages['values']['root']['rows']['values'].append({'values':[]})
+            for index, task in enumerate(tasks):
+                stage_id = 'stage%d' % index
+                stages['values']['root']['rows']['values'][0]['values'].append(stage_id)
+                stages['values'][stage_id] = {'action': {'args': {'values': {'TaskID': task}},'name':'task','timeout':3000},'name':stage_id}
+        payload = {'key': {
+                        'id': cc_id
+                        },
+                        'change': {
+                                'name': name,
+                                'rootStageId': 'root',
+                                'notes': 'randomString',
+                                'stages': stages
+                        }
+        }
+        url = '/api/resources/changecontrol/v1/ChangeControlConfig'
+        # For on-prem check the version as it is only supported from 2021.2.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion >= 6.0:
+                self.log.debug('v6 ' + str(url) + ' ' + str(payload))
+                response = self.clnt.post(url, data=payload, timeout=self.request_timeout)
+                return response
+            elif self.clnt.apiversion < 6.0:
+                    self.log.warning('change_control_create_for_tasks: Change Control Resource APIs are supported'
+                                     ' from 2021.2.0 or newer.')
+                    return None
+        else:
+            return self.clnt.post(url, data=payload, timeout=self.request_timeout)
+
+    def change_control_start(self, cc_id, notes=""):
+        ''' Start a Change Control using Resource APIs.
+            Supported versions: CVP 2021.2.0 or newer and CVaaS.
+            Args:
+                cc_id (string): The ID for the new change control.
```

Comment:
Trying to start a non-existing CC gives the error message: `{"code":9, "message":"not approved"}`
```
print(clnt_imp.api.change_control_start('non-existing-cc-id', notes="my_note"))
```
Result: 
```
---------------------------------------------------------------------------
CvpRequestError                           Traceback (most recent call last)
/var/folders/hd/3g0v07bx6vd9n3tn3ztkc5900000gn/T/ipykernel_20210/1392912492.py in <module>
      1 # Non-existing CC
----> 2 print(clnt_imp.api.change_control_start('non-existing-cc-id', notes="my_note"))

~/Documents/tmp/cvprac/cvprac/cvp_api.py in change_control_start(self, cc_id, notes)
   3623             if self.clnt.apiversion >= 6.0:
   3624                 self.log.debug('v6 ' + str(url) + ' ' + str(payload))
-> 3625                 response = self.clnt.post(url, data=payload, timeout=self.request_timeout)
   3626                 return response
   3627             elif self.clnt.apiversion < 6.0:

~/Documents/tmp/cvprac/cvprac/cvp_client.py in post(self, url, data, files, timeout)
    880                     re-instantiate.
    881         '''
--> 882         return self._make_request('POST', url, timeout, data=data, files=files)
    883 
    884     def delete(self, url, data=None, timeout=30):

~/Documents/tmp/cvprac/cvprac/cvp_client.py in _make_request(self, req_type, url, timeout, data, files)
    615                 full_url = self.url_prefix + url
    616             try:
--> 617                 response = self._send_request(req_type, full_url, timeout,
    618                                               data, files)
    619             except CvpApiError as error:

~/Documents/tmp/cvprac/cvprac/cvp_client.py in _send_request(self, req_type, full_url, timeout, data, files)
    767 
    768             try:
--> 769                 self._is_good_response(response, '%s: %s ' %
    770                                        (req_type, full_url))
    771             except CvpSessionLogOutError as error:

~/Documents/tmp/cvprac/cvprac/cvp_client.py in _is_good_response(self, response, prefix)
    415                                                       response.text)
    416                 self.log.error(msg)
--> 417                 raise CvpRequestError(msg)
    418 
    419         if 'LOG OUT MESSAGE' in response.text:

CvpRequestError: POST: https://<IP>:443/api/resources/changecontrol/v1/ChangeControlConfig : Request Error: Bad Request - {"code":9, "message":"not approved"}
```

Not sure if we can improve that, otherwise I think that's not really a big deal to leave it like that. 

I got the same when trying to stop a non-existing CC: 

```
# Stopping the CC
print(clnt_imp.api.change_control_stop('non-existing-cc-id', notes="my_note2"))
```
Result: 
```
CvpRequestError: POST: https://<IP>:443/api/resources/changecontrol/v1/ChangeControlConfig : Request Error: Bad Request - {"code":9, "message":"not approved"}
```

Reply:
This to me seems like a CVP issue? I think CVP should return a more descriptive error. Essentially something like "CC does not exist"

Reply:
filed a bug on CV side to correct the error message